### PR TITLE
Add RulesConfigProtocolFormatter

### PR DESCRIPTION
**Pull request by rita-sonarsource**
Part of #2537

### CODE REVIEW

File: src/CFamily/Analysis/RulesConfigProtocolFormatter.cs [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+        public Dictionary<string, string> RuleParameters { get; }
+
+        public RuleConfigProtocolFormat(string qualityProfile, Dictionary<string, string> ruleParameters)
+        {
+            QualityProfile = qualityProfile;
+            RuleParameters = ruleParameters;
+        }
+    }
+
+    public class RulesConfigProtocolFormatter : IRulesConfigProtocolFormatter
```
Comment (rita-sonarsource) [inline src/CFamily/Analysis/RulesConfigProtocolFormatter.cs:51]: The code is shared between cmake and vcx, so I created a helper class.

File: src/Integration.Vsix/CFamily/CMake/CompilationDatabaseRequest.cs [hunk 2]
```diff
@@ -89,11 +99,29 @@ public void WriteRequest(BinaryWriter writer)
...
+            if (!Context.AnalyzerOptions?.CreatePreCompiledHeaders ?? true)
+            {
+                var rulesProtocolFormat = rulesConfigProtocolFormatter.Format(Context.RulesConfiguration);
+
+                WriteQualityProfile(writer, rulesProtocolFormat.QualityProfile);
+                WriteRuleSettings(writer, rulesProtocolFormat.RuleParameters);
+            }
+            else
+            {
+                WriteQualityProfile(writer, "");
```
Comment (rita-sonarsource) [inline src/Integration.Vsix/CFamily/CMake/CompilationDatabaseRequest.cs:121]: I believe we don't need this, I tested the pch generation and it seems fine without it. Will ask the cfamily.

File: src/Integration.Vsix/CFamily/CMake/CompilationDatabaseRequest.cs [hunk 3]
```diff
@@ -33,10 +34,19 @@ namespace SonarLint.VisualStudio.Integration.Vsix.CFamily.CMake
     internal class CompilationDatabaseRequest : IRequest
```
Comment (rita-sonarsource) [inline src/Integration.Vsix/CFamily/CMake/CompilationDatabaseRequest.cs:34]: I haven't changed the tests yet, so they're using the inner class as an integration.
I will be done with the vcx request's tests and then adopt CompilationDatabaseRequestTests.

File: src/CFamily/Analysis/RulesConfigProtocolFormatter.cs [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+            var ruleParameters = GetRuleParameters(rulesConfig);
+
+            return new RuleConfigProtocolFormat(qualityProfile, ruleParameters);
+        }
+
+        private static Dictionary<string, string> GetRuleParameters(ICFamilyRulesConfig rulesConfiguration)
+        {
+            var ruleParameters = new Dictionary<string, string>();
+            
+            foreach (var ruleKey in rulesConfiguration.AllPartialRuleKeys)
```
Comment (duncanp-sonar) [inline src/CFamily/Analysis/RulesConfigProtocolFormatter.cs:70]: If I remember correctly, the VCX version used `AllPartialRuleKeys`, and in the CMake version I used `AllActiveRuleKeys`. There didn't seem to be much point in passing parameters for rules that won't be executed.

File: src/Integration.Vsix/CFamily/CMake/CompilationDatabaseRequest.cs (hunk 2 above)
Comment (duncanp-sonar) [reply src/Integration.Vsix/CFamily/CMake/CompilationDatabaseRequest.cs:121]: It's listed as an essential field in the CMake integration spec, and their code suggests they will error without it:
```
llvm::Optional<DriverError> MSVCDriver::setCompilationIndependentOpt(AnalyzerInput& analyzerInput) const {
...
  if (auto errorOrOpt = getOptionOrError("QualityProfile")) {
...
}
```

It's possible they are returning an error message and we are ignoring it.

**Review by duncanp-sonar — APPROVED** [review]

File: src/CFamily/Analysis/RulesConfigProtocolFormatter.cs (hunk 4 above)
Comment (rita-sonarsource) [reply src/CFamily/Analysis/RulesConfigProtocolFormatter.cs:70]: Good point, I'll align that.

**Pull request merged**